crypto/dilithium/p_dilithium3_asn1.c

```diff
@@ -0,0 +1,206 @@
+  // See https://datatracker.ietf.org/doc/draft-massimo-lamps-pq-sig-certificates/00/ section 4.
+  // The parameters must be omitted.
+  if (CBS_len(params) != 0) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_DECODE_ERROR);
+    return 0;
+  }
+
+  return dilithium3_set_pub_raw(out, CBS_data(key), CBS_len(key));
+}
+
+static int dilithium3_pub_encode(CBB *out, const EVP_PKEY *pkey) {
+  const DILITHIUM3_KEY *key = pkey->pkey.ptr;
+
+  // See https://datatracker.ietf.org/doc/draft-massimo-lamps-pq-sig-certificates/00/ section 4.
+  CBB spki, algorithm, oid, key_bitstring;
+  if (!CBB_add_asn1(out, &spki, CBS_ASN1_SEQUENCE) ||
+      !CBB_add_asn1(&spki, &algorithm, CBS_ASN1_SEQUENCE) ||
+      !CBB_add_asn1(&algorithm, &oid, CBS_ASN1_OBJECT) ||
+      !CBB_add_bytes(&oid, dilithium3_asn1_meth.oid, dilithium3_asn1_meth.oid_len) ||
+      !CBB_add_asn1(&spki, &key_bitstring, CBS_ASN1_BITSTRING) ||
+      !CBB_add_u8(&key_bitstring, 0 /* padding */) ||
+      !CBB_add_bytes(&key_bitstring, key->pub, DILITHIUM3_PUBLIC_KEY_BYTES) ||
+      !CBB_flush(out)) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_ENCODE_ERROR);
+    return 0;
+  }
+
+  return 1;
+}
+
+static int dilithium3_pub_cmp(const EVP_PKEY *a, const EVP_PKEY *b) {
+  const DILITHIUM3_KEY *a_key = a->pkey.ptr;
+  const DILITHIUM3_KEY *b_key = b->pkey.ptr;
+  return OPENSSL_memcmp(a_key->pub, b_key->pub, DILITHIUM3_PUBLIC_KEY_BYTES) == 0;
+}
+
+static int dilithium3_priv_decode(EVP_PKEY *out, CBS *params, CBS *key) {
+  // See https://datatracker.ietf.org/doc/draft-massimo-lamps-pq-sig-certificates/00/ section 5.
+
+  // The parameters must be omitted.
+  if (CBS_len(params) != 0) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_DECODE_ERROR);
+    return 0;
+  }
+
+  return dilithium3_set_priv_raw(out, CBS_data(key), CBS_len(key));
+}
+
+static int dilithium3_priv_encode(CBB *out, const EVP_PKEY *pkey) {
+  DILITHIUM3_KEY *key = pkey->pkey.ptr;
+  if (!key->has_private) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NOT_A_PRIVATE_KEY);
+    return 0;
+  }
+  // See https://datatracker.ietf.org/doc/draft-massimo-lamps-pq-sig-certificates/00/ section 5.
+  CBB pkcs8, algorithm, oid, private_key;
+  if (!CBB_add_asn1(out, &pkcs8, CBS_ASN1_SEQUENCE) ||
+      !CBB_add_asn1_uint64(&pkcs8, 0 /* version */) ||
+      !CBB_add_asn1(&pkcs8, &algorithm, CBS_ASN1_SEQUENCE) ||
+      !CBB_add_asn1(&algorithm, &oid, CBS_ASN1_OBJECT) ||
+      !CBB_add_bytes(&oid, dilithium3_asn1_meth.oid, dilithium3_asn1_meth.oid_len) ||
+      !CBB_add_asn1(&pkcs8, &private_key, CBS_ASN1_BITSTRING) ||
+      !CBB_add_bytes(&private_key, key->priv, DILITHIUM3_PRIVATE_KEY_BYTES) ||
+      !CBB_flush(out)) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_ENCODE_ERROR);
+    return 0;
+  }
+
+  return 1;
+}
+
+static int dilithium3_size(const EVP_PKEY *pkey) {
+  return DILITHIUM3_PUBLIC_KEY_BYTES + DILITHIUM3_PRIVATE_KEY_BYTES;
+}
+
+static int dilithium3_bits(const EVP_PKEY *pkey) {
+  return 8 * (DILITHIUM3_PUBLIC_KEY_BYTES + DILITHIUM3_PRIVATE_KEY_BYTES);
```

Comment:
What is your rationale for this? The comment in evp.h for EVP_PKEY_bits is hard to map to dilithium3_bits.

Reply:
Agreed, I got it wrong here. Looking at [OQS](https://github.com/open-quantum-safe/openssl/blob/OQS-OpenSSL_1_1_1-stable/crypto/ec/oqs_meth.c#L1082) they use the length of the public key - so i've changed to be consistent with this.